### PR TITLE
Flux Control(Depth/Canny) + Inpaint

### PR DESCRIPTION
**Pull request by affromero**
# What does this PR do?

The Flux pipeline for image inpainting using Flux-dev-Depth/Canny

```py
import torch
from diffusers import FluxControlInpaintPipeline
from diffusers.models.transformers import FluxTransformer2DModel
from transformers import T5EncoderModel
from diffusers.utils import load_image, make_image_grid
from image_gen_aux import DepthPreprocessor # https://github.com/huggingface/image_gen_aux
from PIL import Image
import numpy as np

pipe = FluxControlInpaintPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-Depth-dev",
    torch_dtype=torch.bfloat16,
)
# use following lines if you have GPU constraints
# ---------------------------------------------------------------
transformer = FluxTransformer2DModel.from_pretrained(
    "sayakpaul/FLUX.1-Depth-dev-nf4", subfolder="transformer", torch_dtype=torch.bfloat16
)
text_encoder_2 = T5EncoderModel.from_pretrained(
    "sayakpaul/FLUX.1-Depth-dev-nf4", subfolder="text_encoder_2", torch_dtype=torch.bfloat16
)
pipe.transformer = transformer
pipe.text_encoder_2 = text_encoder_2
pipe.enable_model_cpu_offload()
# ---------------------------------------------------------------
pipe.to("cuda")

prompt = "a blue robot singing opera with human-like expressions"
image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/robot.png")

head_mask = np.zeros_like(image)
head_mask[65:580,300:642] = 255
mask_image = Image.fromarray(head_mask)

processor = DepthPreprocessor.from_pretrained("LiheYoung/depth-anything-large-hf")
control_image = processor(image)[0].convert("RGB")

output = pipe(
    prompt=prompt,
    image=image,
    control_image=control_image,
    mask_image=mask_image,
    num_inference_steps=30,
    strength=0.9,
    guidance_scale=50.0,
    generator=torch.Generator().manual_seed(42),
).images[0]
make_image_grid([image, control_image, mask_image, output.resize(image.size)], rows=1, cols=4).save("output.png")
```
output_flux_depth_inpaint
 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@sayakpaul @yiyixuxu 

Following [this](https://x.com/RisingSayak/status/1866793529835745541) interaction about contributing [this](https://huggingface.co/afromero/pipeline_flux_control_inpaint) pipeline to main.

### CODE REVIEW

**Review by sayakpaul — COMMENTED** [review]
Thanks for starting this! I have left some comments, LMK if they make sense.

File: src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py [hunk 1]
```diff
@@ -0,0 +1,1110 @@
...
+        prompt_2: Optional[Union[str, List[str]]] = None,
+        image: PipelineImageInput = None,
+        control_image: PipelineImageInput = None,
+        mask_image: PipelineImageInput = None,
+        masked_image_latents: PipelineImageInput = None,
+        height: Optional[int] = None,
+        width: Optional[int] = None,
+        strength: float = 0.6,
+        num_inference_steps: int = 28,
+        timesteps: List[int] = None,
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py:762]:
```suggestion
        sigmas: Optional[List[float]] = None,
```

File: src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py [hunk 2]
```diff
@@ -0,0 +1,1110 @@
...
+                on the amount of noise initially added. When `strength` is 1, added noise is maximum and the denoising
+                process runs for the full number of iterations specified in `num_inference_steps`. A value of 1
+                essentially ignores `image`.
+            num_inference_steps (`int`, *optional*, defaults to 50):
+                The number of denoising steps. More denoising steps usually lead to a higher quality image at the
+                expense of slower inference.
+            timesteps (`List[int]`, *optional*):
+                Custom timesteps to use for the denoising process with schedulers which support a `timesteps` argument
+                in their `set_timesteps` method. If not defined, the default behavior when `num_inference_steps` is
+                passed will be used. Must be in descending order.
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py:826]:
```suggestion
            sigmas (`List[float]`, *optional*):
                Custom sigmas to use for the denoising process with schedulers which support a `sigmas` argument in
                their `set_timesteps` method. If not defined, the default behavior when `num_inference_steps` is passed
                will be used.
```

File: src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py [hunk 3]
```diff
@@ -0,0 +1,1110 @@
...
+                device,
+                generator,
+            )
+            masked_image_latents = torch.cat((masked_image_latents, mask), dim=-1)
+
+        init_image = self.image_processor.preprocess(image, height=height, width=width)
+        init_image = init_image.to(dtype=torch.float32)
+
+        # 4.Prepare timesteps
+        sigmas = np.linspace(1.0, 1 / num_inference_steps, num_inference_steps)
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py:957]:
```suggestion
        sigmas = np.linspace(1.0, 1 / num_inference_steps, num_inference_steps) if sigmas is None else sigmas
```

**Comment by affromero** [comment]
I believe I've addressed all your comments, @sayakpaul @hlky. Please let me know if there's anything I might have missed.

File: src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py [hunk 4]
```diff
@@ -0,0 +1,1111 @@
...
+        height = 2 * (int(height) // (vae_scale_factor * 2))
+        width = 2 * (int(width) // (vae_scale_factor * 2))
+
+        latents = latents.view(batch_size, height // 2, width // 2, channels // 4, 2, 2)
+        latents = latents.permute(0, 3, 1, 4, 2, 5)
+
+        latents = latents.reshape(batch_size, channels // (2 * 2), height, width)
+
+        return latents
+
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py:576]:
```suggestion

    def enable_vae_slicing(self):
        r"""
        Enable sliced VAE decoding. When this option is enabled, the VAE will split the input tensor in slices to
        compute decoding in several steps. This is useful to save some memory and allow larger batch sizes.
        """
        self.vae.enable_slicing()

    def disable_vae_slicing(self):
        r"""
        Disable sliced VAE decoding. If `enable_vae_slicing` was previously enabled, this method will go back to
        computing decoding in one step.
        """
        self.vae.disable_slicing()

    def enable_vae_tiling(self):
        r"""
        Enable tiled VAE decoding. When this option is enabled, the VAE will split the input tensor into tiles to
        compute decoding and encoding in several steps. This is useful for saving a large amount of memory and to allow
        processing larger images.
        """
        self.vae.enable_tiling()

    def disable_vae_tiling(self):
        r"""
        Disable tiled VAE decoding. If `enable_vae_tiling` was previously enabled, this method will go back to
        computing decoding in one step.
        """
        self.vae.disable_tiling()

```

File: src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py [hunk 5]
```diff
@@ -0,0 +1,1111 @@
...
+        control_image = processor(image)[0].convert("RGB")
+
+        output = pipe(
+            prompt=prompt,
+            image=image,
+            control_image=control_image,
+            mask_image=mask_image,
+            num_inference_steps=30,
+            strength=0.9,
+            guidance_scale=50.0,
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py:106]: `guidance_scale=50.0` seems very high?

**Review by hlky — APPROVED** [review]
Left a suggestion and a couple comments but good to go on my end. Thanks!

File: src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py (hunk 5 above)
Comment (affromero) [reply src/diffusers/pipelines/flux/pipeline_flux_control_inpaint.py:106]: uh! that was a last test I was doing to be able to somehow enable `test_flux_different_prompts`, but could not make it work. That is the reason I disabled it `test_flux_different_prompts` in the tests. Maybe I am missing something there. 
Will fix this.

**Comment by asomoza** [comment]
it works ok, I still have to see if it works for some more real use cases, but it produces the expected results.

|a yellow martian|
|---|
|output|

Funny that it generated this "unknown logo" with this example  ^^

image

**Review by asomoza — APPROVED** [review]

**Comment by hlky** [comment]
@affromero here's a couple example PRs for documentation https://github.com/huggingface/diffusers/pull/10021/files https://github.com/huggingface/diffusers/pull/10131/files files under docs/. Ping @stevhliu if you need help with the docs or when they're ready for review.

**Comment by affromero** [comment]
I think this should be good, but not 100% sure @stevhliu, let me know what I can tweak or do next.

**Comment by sayakpaul** [comment]
Will let @hlky take care of the final merging.

**Comment by hlky** [comment]
Failing test is unrelated, good to go, thanks @affromero!

**Comment by vladmandic** [comment]
please update `src/diffusers/pipelines/auto_pipeline.py` as well.

@yiyixuxu @sayakpaul @asomoza can we make this somehow part of the mandatory template? 90% of the time when any img2img or inpaint pipeline is added, autopipeline mapping is NOT updated.

**Pull request merged**